pennylane/spin/spin_hamiltonian.py

```diff
@@ -0,0 +1,88 @@
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains functions to create different templates of spin Hamiltonians.
+"""
+
+from pennylane import X, Z, math
+
+from .lattice import _generate_lattice
+
+# pylint: disable=too-many-arguments
+
+
+def transverse_ising(
+    lattice, n_cells, coupling=None, h=1.0, boundary_condition=False, neighbour_order=1
+):
+    r"""Generates the transverse field Ising model on a lattice.
+    The Hamiltonian is represented as:
+    .. math::
+
+        \hat{H} =  -J \sum_{<i,j>} \sigma_i^{z} \sigma_j^{z} - h\sum{i} \sigma_{i}^{x}
+
+    where J is the coupling defined for the Hamiltonian, h is the strength of transverse
+    magnetic field and i,j represent the indices for neighbouring spins.
+
+    Args:
+       lattice (str): Shape of the lattice. Input Values can be ``'Chain'``, ``'Square'``, ``'Rectangle'``, ``'Honeycomb'``, ``'Triangle'``, or ``'Kagome'``.
+       n_cells (list[int]): Number of cells in each direction of the grid.
+       coupling (List[float] or List[math.array[float]]): Coupling between spins, it can be a list of length equal to neighbour_order or a 2D array of shape number of spins * number of spins. Default value is [1.0].
+       h (float): Value of external magnetic field. Default is 1.0.
+       boundary_condition (bool or list[bool]): Defines boundary conditions, False for open boundary condition, each element represents the axis for lattice. It defaults to False.
+       neighbour_order (int): Range of neighbours a spin interacts with. Default is 1.
+
+    Returns:
+       pennylane.LinearCombination: Hamiltonian for the transverse-field ising model.
+
+    **Example**
+    >>> lattice = "Square"
+    >>> n_cells = [2,2]
+    >>> J = 0.5
+    >>> h = 0.1
+    >>> spin_ham = transverse_ising(lattice, n_cells, coupling=J, h=h)
+    >>> print(spin_ham)
+    -0.5 * (Z(0) @ Z(1))
+    + -0.5 * (Z(0) @ Z(2))
+    + -0.5 * (Z(1) @ Z(3))
+    + -0.5 * (Z(2) @ Z(3))
+    + -0.1 * X(0)
+    + -0.1 * X(1)
+    + -0.1 * X(2)
+    + -0.1 * X(3)
+
+    """
+    lattice = _generate_lattice(lattice, n_cells, boundary_condition, neighbour_order)
+    if coupling is None:
+        coupling = [1.0]
+    coupling = math.asarray(coupling)
+    hamiltonian = 0.0
+    print(coupling.shape)
+    if coupling.shape not in [(neighbour_order,), (lattice.n_sites, lattice.n_sites)]:
+        raise ValueError(
+            f"Coupling shape should be equal to {neighbour_order} or {lattice.n_sites}x{lattice.n_sites}"
+        )
+
+    if coupling.shape == (neighbour_order,):
+        for edge in lattice.edges:
+            i, j, order = edge[0], edge[1], edge[2]
+            hamiltonian += -coupling[order] * (Z(i) @ Z(j))
```

Comment:
So there's no further `coupling[order][i, j]` support? 🤔

Reply:
There are two cases, I have considered here, one is that the user provides the coupling as a constant for each different neighbour order or as a matrix of size num of spins *num of spins. coupling[order][i,j] will be included in the second one, as it iterates over all the edges and gets the coupling value for any two spins that are there in the edges.